libr/core/cmd_api.c

```diff
@@ -355,6 +355,157 @@ R_API RCmdStatus r_cmd_call_parsed_args(RCmd *cmd, RCmdParsedArgs *args) {
+	}
+	return max_len;
+}
+
+static void print_child_help(RStrBuf *sb, RCmdDesc *cd, size_t max_len) {
+	size_t str_len = strlen (cd->name) + strlen0 (cd->help.args_str);
+	size_t padding = str_len < max_len ? max_len - str_len : 0;
+	const char *cd_args_str = cd->help.args_str ? cd->help.args_str : "";
+	const char *cd_summary = cd->help.summary ? cd->help.summary : "";
+	r_strbuf_appendf (sb, "| %s %s %*s# %s\n", cd->name, cd_args_str, padding, "", cd_summary);
+}
+
+static char *inner_get_help(RCmd *cmd, RCmdDesc *cd) {
+	if (!cd->help.args_str || !cd->help.summary) {
+		return NULL;
+	}
+
+	RStrBuf *sb = r_strbuf_new (NULL);
+	fill_usage_strbuf (sb, cd);
+
+	void **it_cd;
+	size_t max_len = 0;
+
+	if (cd->d.argv_data.cb) {
+		max_len = update_max_len (cd, max_len);
+	}
+	r_cmd_desc_children_foreach (cd, it_cd) {
+		RCmdDesc *child = *(RCmdDesc **)it_cd;
+		max_len = update_max_len (child, max_len);
+	}
+
+	if (cd->d.argv_data.cb) {
+		print_child_help (sb, cd, max_len);
+	}
+	r_cmd_desc_children_foreach (cd, it_cd) {
+		RCmdDesc *child = *(RCmdDesc **)it_cd;
+		print_child_help (sb, child, max_len);
+	}
+	return r_strbuf_drain (sb);
+}
+
+static char *argv_get_help(RCmd *cmd, RCmdDesc *cd, RCmdParsedArgs *a, size_t detail) {
+	if (!cd->help.args_str || !cd->help.summary) {
+		return NULL;
+	}
+
+	RStrBuf *sb = r_strbuf_new (NULL);
+	const char **e;
+
+	fill_usage_strbuf (sb, cd);
+
+	switch (detail) {
+	case 1:
+		return r_strbuf_drain (sb);
+	case 2:
+		if (cd->help.description) {
+			r_strbuf_appendf (sb, "\n%s\n", cd->help.description);
+		}
+		if (cd->help.examples) {
+			r_strbuf_append (sb, "\nExamples:\n");
+			for (e = cd->help.examples; *e; e += 2) {
+				r_return_val_if_fail (*(e + 1), NULL);
+				r_strbuf_appendf (sb, "| %s # %s\n", *e, *(e + 1));
+			}
+		}
+		return r_strbuf_drain (sb);
+	default:
+		r_strbuf_free (sb);
+		return NULL;
+	}
+}
+
+static char *oldinput_get_help(RCmd *cmd, RCmdDesc *cd, RCmdParsedArgs *a) {
+	const char *s = NULL;
+	r_cons_push ();
+	RCmdStatus status = r_cmd_call_parsed_args (cmd, a);
+	if (status == R_CMD_STATUS_OK) {
+		s = r_cons_get_buffer ();
+	}
+	char *res = strdup (s? s: "");
+	r_cons_pop ();
+	return res;
+}
+
+R_API char *r_cmd_get_help(RCmd *cmd, RCmdParsedArgs *args) {
+	char *cmdid = strdup (r_cmd_parsed_args_cmd (args));
+	char *cmdid_p = cmdid + strlen (cmdid) - 1;
+	size_t detail = 0;
+	while (cmdid_p >= cmdid && *cmdid_p == '?') {
+		*cmdid_p = '\0';
+		cmdid_p--;
+		detail++;
+	}
+
+	if (detail == 0) {
+		// there should be at least one `?`
+		free (cmdid);
+		return NULL;
+	}
+
+	RCmdDesc *cd = cmdid_p >= cmdid? r_cmd_get_desc (cmd, cmdid): r_cmd_get_root (cmd);
+	free (cmdid);
+	if (!cd) {
+		return NULL;
+	}
+
+	switch (cd->type) {
+	case R_CMD_DESC_TYPE_ARGV:
+		if (detail == 1 && !r_pvector_empty (&cd->children)) {
+			if (args->argc > 1) {
+				return NULL;
+			}
+			return inner_get_help (cmd, cd);
+		}
+		return argv_get_help (cmd, cd, args, detail);
```

Comment:
I'd say that "cmd?" should include the description and examples if children list is empty.

Reply:
Yeah the only "problem" is that e.g. `wv?` would display the list of subcommands, so in those cases you would need to use `wv??` while for others you would use `wv8?`. But yeah, I agree with you though I'm not sure how to solve this issue.